### PR TITLE
Simplify expressions with `NOT` clause

### LINKED ISSUE #2957: simplify not clause
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

simplify not clause if possible. 

For example: 
not (A > B) ==> (A <= B)
not (A or B) ==> (not A) and (not B)
not (A is not null) ==> A is null
not (A not in (..)) ===> A in (..)
....

**Describe the solution you'd like**
Above.

**Describe alternatives you've considered**
None

**Additional context**
None

### PR DESCRIPTION
**Pull request by AssHero**
# Which issue does this PR close?

Closes #2957 

 # Rationale for this change
negate a Not clause

For BinaryExpr, use the negator of op instead.
    not ( A > B) ===> (A <= B)
 For BoolExpr, not (A and B) ===> (not A) or (not B)
     not (A or B) ===> (not A) and (not B)
     not (not A) ===> A
 For NullExpr, not (A is not null) ===> A is null
     not (A is null) ===> A is not null
 For InList, not (A not in (..)) ===> A in (..)
     not (A in (..)) ===> A not in (..)
 For Between, not (A between B and C) ===> (A not between B and C)
     not (A not between B and C) ===> (A between B and C)
 For others, use Not clause

# What changes are included in this PR?
Add rules to simplify not clause in datafusion/optimizer/src/simplify_expressions.rs

### CODE REVIEW

File: datafusion/core/tests/sql/joins.rs [hunk 1]
```diff
@@ -1567,7 +1567,7 @@ async fn reduce_right_join_2() -> Result<()> {
     let expected = vec![
         "Explain [plan_type:Utf8, plan:Utf8]",
         "  Projection: #t1.t1_id, #t1.t1_name, #t1.t1_int, #t2.t2_id, #t2.t2_name, #t2.t2_int [t1_id:UInt32;N, t1_name:Utf8;N, t1_int:UInt32;N, t2_id:UInt32;N, t2_name:Utf8;N, t2_int:UInt32;N]",
-        "    Filter: NOT #t1.t1_int = #t2.t2_int [t1_id:UInt32;N, t1_name:Utf8;N, t1_int:UInt32;N, t2_id:UInt32;N, t2_name:Utf8;N, t2_int:UInt32;N]",
+        "    Filter: #t1.t1_int != #t2.t2_int [t1_id:UInt32;N, t1_name:Utf8;N, t1_int:UInt32;N, t2_id:UInt32;N, t2_name:Utf8;N, t2_int:UInt32;N]",
```
Comment (alamb) [inline datafusion/core/tests/sql/joins.rs:1570]: 👍

File: datafusion/optimizer/src/simplify_expressions.rs [hunk 2]
```diff
@@ -185,6 +181,97 @@ fn as_bool_lit(expr: Expr) -> Option<bool> {
...
+///     not (A in (..)) ===> A not in (..)
+/// For Between, not (A between B and C) ===> (A not between B and C)
+///     not (A not between B and C) ===> (A between B and C)
+/// For others, use Not clause
+fn negate_clause(expr: Expr) -> Expr {
+    match expr {
+        Expr::BinaryExpr { left, op, right } => {
+            match op {
+                // not (A = B) ===> (A <> B)
+                Operator::Eq => binary_expr(*left, Operator::NotEq, *right),
```
Comment (alamb) [inline datafusion/optimizer/src/simplify_expressions.rs:203]: I don't know if it matters, but `binary_expr` may cause a new allocation via `Box` -- I wonder if this code could be written to just pick a new op and reassmble the expr, like

```rust
        Expr::BinaryExpr { left, op, right } => Expr::BinaryExpr { left, op: negate_op(op), right }
```
not necessary, just something that I thought of

File: datafusion/optimizer/src/simplify_expressions.rs [hunk 3]
```diff
@@ -185,6 +181,97 @@ fn as_bool_lit(expr: Expr) -> Option<bool> {
...
+                    binary_expr(*left, Operator::IsDistinctFrom, *right)
+                }
+                // not (A and B) ===> (not A) or (not B)
+                Operator::And => {
+                    let left = negate_clause(*left);
+                    let right = negate_clause(*right);
+
+                    or(left, right)
+                }
+                // not (A or B) ===> (not A) and (not B)
```
Comment (alamb) [inline datafusion/optimizer/src/simplify_expressions.rs:233]: nice

File: datafusion/optimizer/src/simplify_expressions.rs [hunk 4]
```diff
@@ -185,6 +181,97 @@ fn as_bool_lit(expr: Expr) -> Option<bool> {
...
+            negated,
+            low,
+            high,
+        } => Expr::Between {
+            expr,
+            negated: !negated,
+            low,
+            high,
+        },
+        // use not clause
```
Comment (alamb) [inline datafusion/optimizer/src/simplify_expressions.rs:270]: 👍  I reviewed this logic carefully

File: datafusion/optimizer/src/simplify_expressions.rs [hunk 5]
```diff
@@ -905,7 +987,7 @@ mod tests {
             Operator::And,
             Expr::not(col("c2").gt(lit(5))),
         );
-        let expected = expr.clone();
+        let expected = col("c2").gt(lit(5)).and(col("c2").lt_eq(lit(5)));
```
Comment (alamb) [inline datafusion/optimizer/src/simplify_expressions.rs:990]: Maybe can we update the comment above as well

from 
```rust
        // (c > 5) AND !(c > 5) -- can't remove
```

to 
```rust
        // (c > 5) AND !(c > 5) -- > (c > 5) AND (c <= 5)
```

File: datafusion/optimizer/src/simplify_expressions.rs [hunk 6]
```diff
@@ -1434,7 +1516,7 @@ mod tests {
                 ],
                 else_expr: Some(Box::new(lit(true))),
             })),
-            col("c1").or(col("c1").or(col("c2")).not())
+            col("c1").or(col("c1").not().and(col("c2").not()))
```
Comment (alamb) [inline datafusion/optimizer/src/simplify_expressions.rs:1519]: Can you please update the comments here as well?

File: datafusion/optimizer/src/simplify_expressions.rs [hunk 7]
```diff
@@ -1453,7 +1535,7 @@ mod tests {
...
```
Comment (alamb) [inline datafusion/optimizer/src/simplify_expressions.rs:1538]: And one more place to update the comments?

**Review by alamb — APPROVED** [review]
Thank you @AssHero -- this PR was a joy to read ❤️

File: datafusion/optimizer/src/simplify_expressions.rs (hunk 6 above)
Comment (AssHero) [reply datafusion/optimizer/src/simplify_expressions.rs:1519]: Yes, I'll update the comments! Thanks!

File: datafusion/optimizer/src/simplify_expressions.rs (hunk 2 above)
Comment (AssHero) [reply datafusion/optimizer/src/simplify_expressions.rs:203]: I think it's better than new allocation.  Let me fix this. Thanks!

**Comment by AssHero** [comment]
Refine the code according to the comments.

**Review by liukun4515 — APPROVED** [review]
LGTM

**Comment by alamb** [comment]
Thanks @AssHero  and @liukun4515 !

**Pull request merged**